flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/RestartPipelinedRegionFailoverStrategy.java

```diff
@@ -176,15 +183,24 @@ public RestartPipelinedRegionFailoverStrategy(
 
             // if a needed input result partition is not available, its producer region is involved
             for (SchedulingExecutionVertex vertex : regionToRestart.getVertices()) {
-                for (SchedulingResultPartition consumedPartition : vertex.getConsumedResults()) {
-                    if (!resultPartitionAvailabilityChecker.isAvailable(
-                            consumedPartition.getId())) {
-                        SchedulingPipelinedRegion producerRegion =
-                                topology.getPipelinedRegionOfVertex(
-                                        consumedPartition.getProducer().getId());
-                        if (!visitedRegions.contains(producerRegion)) {
-                            visitedRegions.add(producerRegion);
-                            regionsToVisit.add(producerRegion);
+                for (ConsumedPartitionGroup consumedPartitionGroup :
+                        vertex.getConsumedPartitionGroups()) {
+                    if (!visitedConsumedResultGroups.contains(consumedPartitionGroup)) {
+                        visitedConsumedResultGroups.add(consumedPartitionGroup);
+                        for (IntermediateResultPartitionID consumedPartitionId :
+                                consumedPartitionGroup) {
+                            if (!resultPartitionAvailabilityChecker.isAvailable(
```

Comment:
maybe extract the finding of unvisited partitions out to avoid such deep nested loops? e.g.

```
    private Iterable<IntermediateResultPartitionID> getUnvisitedConsumedPartitions(
            SchedulingPipelinedRegion region,
            Set<ConsumedPartitionGroup> visitedConsumedResultGroups) {

        final List<ConsumedPartitionGroup> unvisitedConsumedPartitionGroups = new ArrayList<>();
        for (SchedulingExecutionVertex vertex : region.getVertices()) {
            for (ConsumedPartitionGroup consumedPartitionGroup :
                    vertex.getConsumedPartitionGroups()) {
                if (!visitedConsumedResultGroups.contains(consumedPartitionGroup)) {
                    continue;
                }

                visitedConsumedResultGroups.add(consumedPartitionGroup);
                unvisitedConsumedPartitionGroups.add(consumedPartitionGroup);
            }
        }

        return () -> IterableUtils.flatMap(unvisitedConsumedPartitionGroups, Function.identity());
    }
```

Reply:
Thank you for proposing this improvement, @zhuzhurk. I've already implemented it in the fix-up commit.